shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/UnitTests.hs

```diff
@@ -172,6 +176,60 @@ testTruncateUnitInterval = testProperty "truncateUnitInterval in [0,1]" $
     let x = intervalValue $ truncateUnitInterval n
      in (x <= 1) && (x >= 0)
 
+newtype VRFNatVal = VRFNatVal Natural
+  deriving (Show)
+
+instance Arbitrary VRFNatVal where
+  arbitrary =
+    VRFNatVal . fromIntegral
+      <$> choose @Integer
+        ( 0,
+          2
+            ^ ( 8
+                  * VRF.sizeOutputVRF
+                    (Proxy @(VRF (ConcreteCrypto ShortHash)))
+              )
+        )
+  shrink (VRFNatVal v) = VRFNatVal <$> shrinkIntegral v
+
+newtype ASC = ASC ActiveSlotCoeff
+  deriving (Show)
+
+instance Arbitrary ASC where
+  arbitrary =
+    ASC
+      . mkActiveSlotCoeff
+      . unsafeMkUnitInterval
+      . fromRational
+      . toRational
+      <$> choose @Double (0.01, 0.5)
+
+newtype StakeProportion = StakeProportion Rational
+  deriving (Show)
+
+instance Arbitrary StakeProportion where
+  arbitrary = StakeProportion . toRational <$> choose @Double (0, 1)
+
+-- | Test @checkLeaderVal@ in 'Shelley.Spec.Ledger.BlockChain'
+testCheckLeaderVal ::
+  forall v.
+  (v ~ VRF (ConcreteCrypto ShortHash)) =>
+  TestTree
+testCheckLeaderVal =
+  testGroup
+    "Test checkLeaderVal calculation"
+    [ testProperty "With a stake of 0, cannot lead" $
```

Comment:
In the QuickCheck course for IOHK, John Hughes cites a test from the old code base as an example of bad testing: the example is the leadership check; the test was checking that you're never elected when you have no stake, always elected when you have all the stake, and a few other things. He says "the one thing we're really interested in here, _we are elected with probability relative to our stake_, is the one thing that's not actually being tested". It's rather unfortunate that we are now reintroducing this very same test again.

Also, shouldn't there be some tests that all this Taylor stuff should at least correspond in some way to the simple calculation using floating point?

Reply:
I did start to write that test, and then convinced myself it wouldn't be terribly useful since we'd effectively just be testing the uniformity of the VRF generator. But it's true that there's use in testing that, _given_ a uniform VRF function, we are elected with probability relative to our stake. So I'll add that test :-)

There are tests for the non-integer calculations in the relevant package - in this case, we have:
```
property_praos_leader_comparison :: IO()
property_praos_leader_comparison =
  qcWithLabel "property σ,p in [0,1]: p < 1 - (1 - f)^σ <=> taylorExpCmp 3 (1/(1 - p)) (-σ · ln (1 - f))"
    10000 prop_LeaderCm
```